### ISSUE TITLE
Zenodo download links in clld_meta contributions table

### ISSUE THREAD

**Issue by HedvigS** [cldf-datasets/clld_meta#4]
is there any chance that clld_meta/contributions.csv could include Zenodo download links?

I.e in addition to ""https://zenodo.org/records/7844558" also "https://zenodo.org/records/7844558/files/grambank/grambank-v1.0.3.zip" :slightly_smiling_face:?

**Comment by johenglisch** [comment]
Sounds useful. I'd have to think a bit on the details because a single zenodo record can have multiple downloadable files.

**Comment by HedvigS** [comment]
I would like to be able to do something like this in R:

```

library(remotes)
install_github("SimonGreenhill/rcldf", dependencies = TRUE, ref = "v1.1.0")
library(rcldf)

Dataset_string <- "Grambank"
Version <- "most recent"

clld_meta_table <- read.delim("https://github.com/cldf-datasets/clld_meta/raw/4d544e6e27ec79fa2a572827981454115e15fef1/cldf/contributions.csv", sep = ",")

clld_meta_table_spec <- clld_meta_table[grepl(pattern = Dataset_string, x = clld_meta_table$Name),]

if(Version == "most recent"){
Version_str <- sort(clld_meta_table_spec$Version, decreasing = T)[1]}else{
Version_str <- Version
}

link <- clld_meta_table_spec[clld_meta_table_spec$Version == Version_str, "Zenodo_Link"]

rcldf::rcldf(link)

```

**Comment by HedvigS** [comment]
> Sounds useful. I'd have to think a bit on the details because a single zenodo record can have multiple downloadable files.

rcldf deals with the whole zip file pretty well so that's what I want :)

**Comment by xrotwang** [comment]
This sounds like data that should go into a `MediaTable` component in the dataset - which would also allow for the one-to-many relationship between datasets and files.

There's another complication, though: In some cases, deposits on Zenodo contain multiple CLDF datasets, e.g. [Liljegren's Hindu Kush data](https://zenodo.org/records/4534221). So the relation between CLDF dataset - i.e. contribution in the meta dataset - and media files is many-to-many.

**Comment by xrotwang** [comment]
And then there's CLDF datasets, where each file has been added separately to the Zenodo deposit, e.g. https://zenodo.org/records/7085903

Thus, there are three types of Zenodo download links that `rcldf` would need to handle:
- "standard", zipped GitHub repos
- zipped GitHub repos containing multiple CLDF datasets
- CLDF dataset files as individual files on Zenodo.

**Comment by xrotwang** [comment]
Considering these complications, it may be simpler, if `rcldf` already used the Zenodo API to get the download URLs for the record and then proceeded like `cldfzenodo` does, see https://github.com/cldf/cldfzenodo/blob/0b2bcd7a8ebac66b29c47ffc05934fbde741a626/src/cldfzenodo/record.py#L212-L240

**Comment by johenglisch** [comment]
Yeah, that's the kind of stuff I was alluding to with ‘thinking a bit on the details’.  For our data sets we usually have the following schema:

 * 1 *Zenodo Concept* per data set (i.e. the ‘latest version’ link)
 * Multiple *Zenodo Records* per concept (i.e. the entries in *contributions.csv*)
 * 1 *zip file* per record containing a copy the whole cldfbench project
 * Multiple *cldf data sets* within the same zip file (like a word list and a dictionary, for example)

But the *1 zip file per record* part is just a thing we do and not guaranteed at all across all of Zenodo.

There could be multiple zip files or there could be no zip files (the latter isn't actually handled by *clld-meta*, yet).

**Comment by xrotwang** [comment]
I think the division of labour between the dataset catalog and `rcldf` is outlined in the [CLDF dataset discovery spec](https://github.com/cldf/cldf/blob/master/extensions/discovery.md#cldf-dataset-discovery).

**Comment by johenglisch** [comment]
>Implementations must provide a function or service which accepts a dataset
>locator as specified above

So, does that mean that one should be able to just feed a DOI + ID into *pycldf* and *rcldf* and they just Do the Right Thing™?  Is that implemented for both of them?

If so, then yeah, I could just see if I can add a column with data set locators somewhere; probably in *datasets.csv* because that seems to be the level these locators operate on.

**Comment by xrotwang** [comment]
> So, does that mean that one should be able to just feed a DOI + ID into _pycldf_ and _rcldf_ and they just Do the Right Thing™? Is that implemented for both of them?

Yes, that's the idea. It's implemented for `pycldf` - provided you have `cldfzenodo` installed. The idea is that reosolvers for other hosting platforms might be installed, too, e.g. for gitlab or other research data repositories.

**Comment by HedvigS** [comment]
I like the idea of a DOI + ID and Do the Right Thing™ ^^! 

For rcldf, we need to run it by Simon and maybe he would also appreciate help from the DLCE scientific programming team in setting it up. 

Shall I start an issue over at rcldf and tag you two?

**Comment by xrotwang** [comment]
From my POV, the simplest workflow would be to actually use the functionality from `pycldf`/`cldfzenodo`:
```shell
$ cldf createdb --download-dir /tmp/ https://doi.org/10.5281/zenodo.7085902 cinwa.sqlite
https://zenodo.org/api/records?allversions=true&q=doi%3A%2210.5281%2Fzenodo.7085902%22
https://zenodo.org/api/records?q=conceptdoi%3A%2210.5281%2Fzenodo.7085902%22
INFO    <cldf:v1.0:Wordlist at /tmp> loaded in cinwa.sqlite
```
and start in R from there :)

**Comment by xrotwang** [comment]
> just one, comparatively small, thing.

This "comparatively small thing" would obviate the need for `rcldf` entirely, adding more flexible CSV reading, reliable BibTeX parsing and CLDF-conformant column access by propertyUrl :)

**Comment by HedvigS** [comment]
> > just one, comparatively small, thing.
> 
> This "comparatively small thing" would obviate the need for `rcldf` entirely, adding more flexible CSV reading, reliable BibTeX parsing and CLDF-conformant column access by propertyUrl :)

I disagree. But, there we are and it's okay. We have different ideas about the user community and their needs and behaviours and how people respond to the existing tools. The product owner decides, I just say how I as a user and employee in the department think.

To be clear "_just one, comparatively small, thing_" referred to the stepping outside of r into python for a handy zenodo link. I think people would rather just hardcode in links and look them up themselves than switch platforms and install a bunch of more things.

Regarding this specific matter, there seem to be these choices:

1) enrich clld_meta datasets.csv with Zenodo download links
2) educate and encourage people to use pycldf more
3) enrich rcldf to be able to Do the Right Thing™ when it gets a Zenodo DOI + ID
4) do nothing and just continue with the status quo which is looking up and keeping track of dataset zenodo links on their own.

I'm starting to lean to (4).

**Comment by johenglisch** [comment]
I've been thinking a little over the weekend.

afaict the challenge with the data locators is that it's not possible to resolve them with a quick call to the Zenodo API because you have to peek *inside* the zip files and and read all of the json files to see if the id part of the locator matches.  So, just Doing the Right Thing™ is far from trivial for pycldf or rcldf.

So, basically I'm starting to lean towards a variant of option 1.  I already have a collection of zip files on my hard drive, and a list of urls where I got them from, which means, clld-meta would be the easiest place to map data locators to download links (at least for data sets on Zenodo).

**Comment by johenglisch** [comment]
>To be clear "just one, comparatively small, thing" referred to the stepping outside of r into python

Come to think of it, does R have a function that just lets you just run a shell command?  Basically something similar to this (just guessing the API here):

    doi <- "https://zenodo.org/records/7085903"
    out_dir <- "./dataset-download"
    status <- system("cldfbench", "zenodo.download", doi, "--directory", out_dir)
    if (status != 0) { print("oh no, download failed") }

Because that would allow you to use non-R components without *technically* stepping out of the platform.  I think that's what a lot of ‘all-in-one solution slowly taking over your entire life‘ environments do (like Emacs, Vim, Lisp systems, etc.).

**Comment by xrotwang** [comment]
The UNIX shell is **the one** platform I urge people **not** to avoid. You can live in R studio all you want, when it comes to reproducibility, e.g. on the cluster, you are stuck with the shell. And if the cluster is the only time you interact with the shell, you will have a much harder time than when using the shell in daily life.

**Comment by HedvigS** [comment]
> The UNIX shell is **the one** platform I urge people **not** to avoid. You can live in R studio all you want, when it comes to reproducibility, e.g. on the cluster, you are stuck with the shell. And if the cluster is the only time you interact with the shell, you will have a much harder time than when using the shell in daily life.

I agree. I use shell all the time. I don't think it matters that much what we three do or think, what matters more is what other users do and the resources that are available to people in our department.

Personally, I think the obstacle would be the python library installations. I don't fear shell, but I am a bit afraid of python environments and library installations.

**Comment by xrotwang** [comment]
> Personally, I think the obstacle would be the python library installations. I don't fear shell, but I am a bit afraid of python environments and library installations.

I can only recommend investing the time to install `cldfzenodo` and try out the snippet I posted above - for comparison.

**Comment by HedvigS** [comment]
> > Personally, I think the obstacle would be the python library installations. I don't fear shell, but I am a bit afraid of python environments and library installations.
> 
> I can only recommend investing the time to install `cldfzenodo` and try out the snippet I posted above - for comparison.

Understood!

If you want more people to do that, I guess the place to add content to is https://github.com/dlce-eva/dlce-eva/tree/main/doc . If you want, I can help you advertise these guides to people inside our department and collaborators.

Johannes has made a suggestion for (1). Is it okay with @xrotwang if @johenglisch adds Zenodo download URLs to clld_meta/datasets.csv as well?

**Comment by xrotwang** [comment]
> If you want more people to do that, I guess the place to add content to is https://github.com/dlce-eva/dlce-eva/tree/main/doc

What more content than
https://github.com/dlce-eva/dlce-eva/blob/main/doc/python.md#virtual-environments
would you need?
